New expression: count_leading_zeros_exprt

tautschnig opened the pull request:
Rather than ad-hoc handling __builtin_clz and __lzcnt (and their
variants) in the C front-end, make counting-leading-zeros available
across the code base.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
New expression: count_leading_zeros_exprt

Rather than ad-hoc handling __builtin_clz and __lzcnt (and their
variants) in the C front-end, make counting-leading-zeros available
across the code base.

kroening commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -2689,27 +2689,22 @@ exprt c_typecheck_baset::do_special_functions(
     {
-      return nil_exprt{};
+      simplify(clz, *this);
     }
```
I wouldn't simplify these up-front -- the simplifier is called by the front-end whenever a constant is needed.

tautschnig replied:
Ah, true, will fix.

kroening commented on src/goto-programs/goto_clean_expr.cpp:
```diff
@@ -429,6 +438,22 @@ void goto_convertt::clean_expr(
+    {
+      clz.zero_permitted(true);
+      const source_locationt source_location = clz.source_location();
+
```
I would go stronger than the nondet: The wording in the gcc docs is that __builtin_clz is undefined when the argument is zero. Assuming they use the meaning of "undefined" as in the ISO C standard, then this is worth an assertion, to be generated in goto_check.

tautschnig replied:
Hmm, true, then I could perhaps demote the `ID_zero_permitted` to `ID_C_zero_permitted`?

kroening replied:
I don't think you need it at all. We don't have anything equivalent for any of the other expressions that are only partially defined.

tautschnig replied:
But we need to distinguish whether we're handling MSVC's `__lzcnt` or `__builtin_clz`. The former is defined for all values, the latter isn't defined for `0`.

kroening replied:
Ah, right. Then I'd make it `ID_zero_permitted`; a comment would be too weak, given that the semantics do differ.

tautschnig replied:
Except they only sort-of do: it seems that in practice it's almost always ok, it's just that GCC doesn't provide a guarantee: https://stackoverflow.com/questions/19527897/how-undefined-are-builtin-ctz0-or-builtin-clz0. How about the solution that I just pushed?

tautschnig committed:
Replace "bounds_check" by ID_C_bounds_check

Replaces the hard-coded string by a dstringt to avoid typos, and also
changes it to a comment: the annotation (communicating from the
front-end to goto_checkt) does not change the semantics of the
expression that has been annotated.

martin-cs reviewed (approved):
Seems reasonable if we want to go that way.

tautschnig committed:
Make lower_popcount a method of popcount_exprt

Expression lowering may be useful well outside the solvers/ context.
Making it a method of the class being lowered also serves as a way of
describing the semantics of the expression.

This is mainly a refactoring, no changes in behaviour are intended.
